File: source/Controller/Controller.cpp
#include "Controller.h"

using namespace Controller;

TheController::TheController(DependencyInjectionManager* dim)
{
    
    this->log = dim->get<ILogger>();

    this->confs = dim->get<Config>();

    if (dim->contains<ThreadPool>())
    {
        this->tasker = dim->get<ThreadPool>();

    }
    else
    {
        log->info("TheController", "No thread pool found. creating a new one");
        this->tasker = new ThreadPool(4);
    }

    if (this->confs == NULL)
        log->error("TheController", "configuration system can't be found in the dependency injection manager");


    db = dim->get<StorageInterface>();

    this->confs->observate("maxTimeWaitingClient", [&](DynamicVar newValue){
        this->maxTimeWaitingClient = newValue.getInt64();
    }, maxTimeWaitingClient);

    srand(Utils::getCurrentTimeMilliseconds());
}

TheController::~TheController(){
    if (db != NULL) delete db;
}

//creates or change a variable
future<void> TheController::setVar(string name, DynamicVar value)
{
    name = "vars."+name;
    //check if name isn't a internal flag var
    if (name.find('_') == 0 || name.find("._") != string::npos)
    {
        log->warning("TheController", "variabls started with underscorn (_) are just for internal flags and can't be setted by clients");
        promise<void> p;
        p.set_value();
        return p.get_future();
    }

    //checks if the variabel is locked
    if (this->getVarInternalFlag(name, "_lock", 0).getInt() == 1)
    {
        log->warning("TheController", "The variable '"+ name + "' is locked and can't be changed by setVar");
        promise<void> p;
        p.set_value();
        return p.get_future();
    }

    return this->internalSetVar(name, value);

}

future<void> TheController::internalSetVar(string name, DynamicVar value)
{
    return tasker->enqueue([&](){

        vector<future<void>> pendingTasks;

        if (name.find('*') != string::npos)
        {
            #ifdef __TESTING__
                Tester::global_test_result = "Invalid setVar parameter. A variable with '*' can't be setted";
            #endif
            log->warning("TheController", "Invalid setVar parameter. A variable with '*' can't be setted");
            //throw "Invalid setVar parameter. A variable with '*' can't be setted";
            return;
        }


        //set the variable
        db->set(name, value.getString());

        notifyVarModification(name, value);
    });
}

future<void> TheController::lockVar(string varName)
{
    

    return tasker->enqueue([&](string vName){
        string finalVName = "vars."+varName + "._lock";
        //ir variable if currently locked, add an observer to it "._lock" property and wait the change of this to 0

        function<bool()> tryLockFunc = [&]()
        {
            //try lock var
            bool lockSucess = false;
            Utils::named_lock("varLockerSystem", [&]{
                if (db->get(finalVName, "0").getString() == "0")
                {
                    //set the var property '._lock' to 1 (use setVar to change the ._lock)
                    db->set(finalVName, "1");
                    lockSucess = db->get(finalVName, "0").getString() == "1";
                }
            });

            return lockSucess;
        };

        //await var be 0 to try another lock
        bool lockedWithSucess = false;
        while (!lockedWithSucess)
        {
            lockedWithSucess = tryLockFunc();

            if (!lockedWithSucess)
                usleep(1000 + rand() % 9000);
        }

        return;


    }, varName);


    //set var can't change a lokced var

}

future<void> TheController::unlockVar(string varName)
{
    return tasker->enqueue([&](){
        varLockerMutex.lock();
        string finalVName = "vars."+varName + "._lock";
        db->set(finalVName, "0");
        varLockerMutex.unlock();
    });

}

//start to observate a variable
void TheController::observeVar(string varName, string clientId, ApiInterface* api)
{
    varName = "vars."+varName;

    Utils::named_lock("vars."+varName + ".observationLock", [&](){

        Controller_ClientHelper client(this->db, clientId, api);
        if (!db->exists(varName + "._observers.byId."+clientId))
        {
            //Utils::named_lock("observingSystem"+varName, [&]() //more memory, lock by each variable
            Utils::named_lock("observingSystem", [&]() //less memory, lock all variables
            {
                int actualVar_observersCount = db->get(varName + "._observers.list.count", 0).getInt();
                db->set(varName + "._observers.list.count", actualVar_observersCount);

                db->set(varName + "._observers.list."+to_string(actualVar_observersCount), clientId);

                db->set(varName + "._observers.byId."+clientId, actualVar_observersCount);

                client.registerNewObservation(varName);
            });
        }
    });
    
}

void TheController::notifyVarModification(string name, DynamicVar value)
{
    /*string varName = "vars."+varName;

    int actualVar_observersCount = db->get(varName + "._observers.list.count", 0).getInt();
    for (int c =0; c < actualVar_observersCount; c++)
    {
        tasker->enqueue([&](string clientId){

        }, db->get(varName + "._observers.list."+to_string(c), "");
    }*/
}

//stop observate variable
void TheController::stopObservingVar(string clientId, string varName)
{
    varName = "vars."+varName;

    Utils::named_lock("vars."+varName + ".observationLock", [&](){

        int actualVar_observersCount = db->get(varName + "._observers.list.count", 0).getInt();

        for (int c = actualVar_observersCount-1; c>=0; c--)
        {
            if (db->get(varName + "._observers.list."+to_string(c), "").getString() == clientId)
            {
                for (int c2 = c; c2> < actualVar_observersCount; c2++)
                {
                    auto currId = db->get(varName + "._observers.list."+to_string(c2+1), "").getString();
                    db->set(varName+"._observers.list."+to_string(c), currId);

                    db->set(varName + "._observers.byId."+currId, c2);

                }

                //remove the last item from the _observers.list
                db->del(varName + "._observers.list."+to_string(actualVar_observersCount-1));
                actualVar_observersCount--;

                //remove the item from _observers.byId
                db->del(varName + "._observers.byId"+clientId);
            }
        }
    });
}

void TheController::apiStarted(ApiInterface *api)
{
    this->apis[api->getApiId()] = api;
}

string TheController::_createUniqueId()
{
    string tmp = to_string(Utils::getCurrentTimeMilliseconds());
    tmp += to_string(rand()) + to_string(rand());
    return Utils::StringToHex(tmp);
}

future<vector<tuple<string, DynamicVar>>> TheController::getVar(string name, DynamicVar defaultValue)
{
    name = "vars."+name;
    return tasker->enqueue([name, defaultValue, this](){

        function <vector<tuple<string, DynamicVar>>(string name, bool childsToo)> readFromDb;
        readFromDb = [&](string nname, bool childsToo)        
        {
            vector<tuple<string, DynamicVar>> result;    
            string value = db->get(nname, "___invalid____").getString();
            if (value != "___invalid___")
                result.push_back(make_tuple( nname,  value));
            
            if (childsToo)
            {
                auto childs = db->getChilds(name);
                for (auto curr: childs)
                {
                    auto tmp = readFromDb(curr, childsToo);                    
                    result.insert(result.begin(), tmp.begin(), tmp.end());
                }
            }

            return result;
        };

        //if variable ends with *, determine just their name
        string name2 = name;
        bool childsToo = false;
        if (name2.find(".*") != string::npos)
        {
            childsToo = true;
            name2 = name2.substr(0, name2.size()-2);
        }

        //load the valures
        auto values = readFromDb(name2, childsToo);
        

        //if nothing was found, add the default value to the result
        if (values.size() == 0)
            values.push_back(make_tuple(name2, defaultValue));

        //return the values

        return values;
    });
}

future<void> TheController::delVar(string varname)
{
    varname = "vars."+varname;

}

future<vector<string>> TheController::getChildsOfVar(string parentName)
{
    parentName = "vars."+parentName;

}


DynamicVar TheController::getVarInternalFlag(string vName, string flagName, DynamicVar defaultValue)
{ 
    if (flagName != "")
    {
        if (flagName[0] != '_');
            flagName = "_"+flagName;
        
        auto flagValue = this->getVar(vName + "."+flagName, defaultValue).get();
        return std::get<1>(flagValue[0]);
    }
}

void TheController::setVarInternalFlag(string vName, string flagName, DynamicVar value)
{
    if (flagName != "")
    {
        if (flagName[0] != '_');
            flagName = "_"+flagName;
        
        this->internalSetVar(vName + "."+flagName, value).get();
    }
}

string TheController::clientConnected(string clientId, ApiInterface* api)
{
    if (clientId == "")
        clientId = _createUniqueId();

    Controller_ClientHelper client = Controller_ClientHelper(db, clientId, api);

    updateClientAboutObservatingVars(client);

    return clientId;
}

void TheController::updateClientAboutObservatingVars(Controller_ClientHelper controller_ClientHelper)
{
    //For all variabels observed by this client, sent the current value
    atomic<bool> checkClientLiveTime = false;
    auto task = Utils::parallel_foreach<string>(controller_ClientHelper.getObservingVars(), [&](string currVarName)
    {
        auto currVarsAndValues = this->getVar(currVarName, "").get();

        //remove the 'vars.' from the begining of the varname

        if (checkClientLiveTime == false)
        {
            if (controller_ClientHelper.notify(currVarsAndValues) != API::ClientSendResult::LIVE)
                checkClientLiveTime = true;
        }
    }, this->tasker);

    task.wait();

    if (checkClientLiveTime)
        this->checkClientLiveTime(controller_ClientHelper);


}


void TheController::checkClientLiveTime(Controller_ClientHelper client)
{
    if (!client.isConnected())
    {
        if (!client.timeSinceLastLiveTime() >= maxTimeWaitingClient)
        {
            deleteClient(client);
        }
    }

}


void deleteClient(Controller_ClientHelper client)
{
    //remove from vars observations

    //delete the client
}

File: sources/controller/controller.go
package controller

import (
	"errors"
	"strconv"
	"strings"
	"time"

	"rtonello/vss/sources/misc"
	"rtonello/vss/sources/misc/confs"
	"rtonello/vss/sources/misc/logger"
	"rtonello/vss/sources/services/apis"
	"rtonello/vss/sources/services/storage"
)

type IController interface {
	// SetVar creates or updates a variable. It returns an error if the variable
	// is internal (starts with '_') or locked.
	SetVar(name string, value misc.DynamicVar) chan error

	// GetVars returns a channel that will receive a GetVarsResult containing
	// the slice of tuples (name, value) and an error if one occurred. Work runs
	// in a background goroutine.
	GetVars(name string, defaultValue misc.DynamicVar) chan GetVarsResult

	DelVar(name string) chan error

	GetChildsOfVar(parentName string) chan []string

	// LockVar tries to acquire a lock for a variable with a timeout in milliseconds.
	LockVar(varName string, timeoutMs uint) chan error

	// UnlockVar releases the lock for a variable.
	UnlockVar(varName string) chan error

	// IsVarLocked checks if a variable is locked.
	IsVarLocked(varName string) chan bool

	// ObserveVar registers a client observation of a variable and sends the
	// current value to the client via the API. If the client is disconnected,
	// the controller will schedule a liveness check.
	ObserveVar(varName, clientId, customIdsAndMetainfo string, api apis.IApi)

	// StopObservingVar removes a client's observation for a variable.
	StopObservingVar(varName, clientId, customIdsAndMetainfo string, api apis.IApi)

	// ApiStarted registers a new API implementation with the controller.
	ApiStarted(api apis.IApi)

	// ClientConnected registers a client and updates it with currently observed vars.
	ClientConnected(clientId string, api apis.IApi) (string, int)

	// GetSystemVersion returns the configured system version string.
	GetSystemVersion() string
}

// GetVarsResult is the returned value from GetVars: Values is the list of
// (name,value) tuples, Err is non-nil when an error occurred.
type GetVarsResult struct {
	Values []misc.Tuple[misc.DynamicVar]
	Err    error
}

// TheController is a Go port of the C++ TheController class. It coordinates
// variable operations and client notifications using ControllerVarHelper and
// ControllerClientHelper.
type TheController struct {
	log                        logger.ILogger
	confs                      confs.IConfs
	db                         storage.IStorage
	apis                       map[string]apis.IApi
	maxTimeWaitingClientSecond int64
	systemVersion              string
	maxKeyLength               int
	maxKeyWordLength           int
	maxValueSize               int
}

// ControllerClientHelperApi is the minimal API used by controller client helper.
// It mirrors the small apis.IApi used in ControllerClientHelper.
// reuse apis.IApi defined in controller_clienthelper.go

// NewController constructs a controller bound to logger, confs and storage.
func NewController(log logger.ILogger, confs confs.IConfs, db storage.IStorage, systemVersion string) *TheController {
	c := &TheController{
		log:                        log,
		confs:                      confs,
		db:                         db,
		apis:                       make(map[string]apis.IApi),
		systemVersion:              systemVersion,
		maxTimeWaitingClientSecond: 12 * 60 * 60, // default like C++	}
	}

	confMaxTimeWaitingClient := confs.Config("maxTimeWaitingClient_seconds").Value()
	if confMaxTimeWaitingClient.GetInt64() > 0 {
		c.maxTimeWaitingClientSecond = confMaxTimeWaitingClient.GetInt64()
	}

	confMaxKeyLength := confs.Config("maxKeyLength").Value()
	c.maxKeyLength = int(confMaxKeyLength.GetInt64())

	confMaxKeyWordLength := confs.Config("maxKeyWordLength").Value()
	c.maxKeyWordLength = int(confMaxKeyWordLength.GetInt64())

	confMaxValueSize := confs.Config("maxValueSize").Value()
	c.maxValueSize = int(confMaxValueSize.GetInt64())

	return c
}

// SetVar creates or updates a variable. It returns an error if the variable
// is internal (starts with '_') or locked.
func (c *TheController) SetVar(name string, value misc.DynamicVar) chan error {
	ch := make(chan error, 1)
	go func() {
		name = misc.GetOnly(name, "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._/\\,.-*")
		if name == "" {
			ch <- errors.New("variable name cannot be empty")
			return
		}
		if name[0] == '_' || containsUnderscoreDot(name) {
			ch <- errors.New("variables started with underscore are for internal use only")
			return
		}

		if len(name) > c.maxKeyLength {
			ch <- errors.New("variable name exceeds maximum length of " + strconv.Itoa(c.maxKeyLength))
			return
		}

		if len(value.GetString()) > c.maxValueSize {
			ch <- errors.New("variable value exceeds maximum size of " + strconv.Itoa(c.maxValueSize))
			return
		}

		if c.maxKeyWordLength > 0 {
			parts := []string{name}
			if strings.Contains(name, ".") {
				parts = strings.Split(name, ".")
			}

			for _, p := range parts {
				if len(p) > c.maxKeyWordLength {
					ch <- errors.New("a part of the variable name exceeds maximum length of " + strconv.Itoa(c.maxKeyWordLength))
					return
				}
			}
		}

		vh := NewControllerVarHelper(c.db)
		vh.SetControlledVarName(name)
		//accept only text valid chars (remove all special chars)

		if vh.IsLocked() {
			ch <- errors.New("variable is locked")
			return
		}

		strValue := value.GetString()
		value.SetString(strValue)

		vh.SetValue(value)
		// async notify
		go c.notifyVarModification(name, value)
		ch <- nil
	}()
	return ch
}

//  future<GetVarResult> TheController::getVars(string name, DynamicVar defaultValue){
//     return tasker->enqueue([&](string namep, DynamicVar defaultValuep){

//         if (namep == "")
//         {
//             log->warning("TheController", Errors::Error_TheVariableNameCannotBeEmpty);
//             return GetVarResult({}, Errors::Error_TheVariableNameCannotBeEmpty);
//         }

//         function <vector<tuple<string, DynamicVar>>(string namep, bool childsToo)> readFromDb;
//         readFromDb = [&](string nname, bool childsToo)
//         {
//             vector<tuple<string, DynamicVar>> result;
//             Controller_VarHelper varHelper(log, db, nname);

//             if (varHelper.valueIsSetInTheDB())
//                 result.push_back(make_tuple( nname,  varHelper.getValue()));

//             if (childsToo)
//             {
//                 nname = nname == "" ? nname : nname + ".";
//                 auto childs = varHelper.getChildsNames();
//                 for (auto curr: childs)
//                 {
//                     auto tmp = readFromDb(nname + curr, childsToo);
//                     result.insert(result.begin(), tmp.begin(), tmp.end());
//                 }
//             }

//             return result;
//         };

//         //if variable ends with *, determine just their name
//         bool childsToo = false;
//         if (namep.find("*") != string::npos)
//         {
//             if (namep.find("*") == namep.size()-1)
//             {
//                 childsToo = true;
//                 auto p = namep.find(".");
//                 if (p != string::npos && p == namep.size()-2)
//                     namep = namep.substr(0, namep.size()-2);
//                 else
//                     namep = namep.substr(0, namep.size()-1);

//                 //if namep ends with '.', remove it (ex: "a.b.c.*" results in a variable called "a.b.c." - with a ending '.', that is not a valid var name)
//                 if (namep.size() > 0 && namep[namep.size()-1] == '.')
//                     namep = namep.substr(0, namep.size()-1);
//             }
//             else
//             {
//                 log->error("TheController", Errors::Error_WildCardCabBeUsedOnlyAtEndOfVarNameForVarGetting);
//                 return GetVarResult(Errors::Error_WildCardCabBeUsedOnlyAtEndOfVarNameForVarGetting, {});
//             }
//         }

//         //load the valures
//         auto values = readFromDb(namep, childsToo);

//         //if nothing was found, add the default value to the result
//         if (values.size() == 0)
//             values.push_back(make_tuple(namep, defaultValuep));

//         //return the values
//         return GetVarResult(Errors::NoError, values);
//     }, name, defaultValue);

// }

// future<Errors::Error> TheController::delVar(string varname)
// {
//     return tasker->enqueue([&](string varnamep)
//     {
//         auto vars = this->getVar(varnamep, "").get().result;
//         if (vars.size() == 0) {
//             return Errors::Error("Error deleting variable '"+varnamep+"'. Maybe the variable doesn't exist");
//         }

//         vector<Errors::Error> errors;
//         for (auto &currVar: vars)
//         {
//             Controller_VarHelper varHelper(log, db, std::get<0>(currVar));
//             if (!varHelper.deleteValueFromDB())
//                 errors.push_back(Errors::Error("Error deleting variable '"+std::get<0>(currVar)+"'. Maybe the variable doesn't exist"));
//             else
//                 notifyVarModification(std::get<0>(currVar), "");
//         }

//         Errors::Error returnErrors = Errors::NoError;

//         if (errors.size() > 0)
//         {
//             returnErrors = "";
//             for (auto &currError: errors)
//                 returnErrors += (returnErrors == "" ? "" : " + ") + currError;
//         }

//         return returnErrors;
//     }, varname);
// }

// future<Errors::ResultWithStatus<vector<string>>> TheController::getChildsOfVar(string parentName)
// {
//     return tasker->enqueue([&](string parentNamep)
//     {
//         if (parentNamep.find('*') != string::npos)
//         {
//             log->error("TheController", Errors::Error_WildcardCanotBeUsesForGetVarChilds);
//             return Errors::ResultWithStatus<vector<string>>(Errors::Error_WildcardCanotBeUsesForGetVarChilds, {});
//         }

//         Controller_VarHelper varHelper(log, db, parentNamep);
//         auto result = varHelper.getChildsNames();
//         return Errors::ResultWithStatus<vector<string>>(Errors::NoError, result);
//     }, parentName);

// }

func containsUnderscoreDot(name string) bool {
	// returns true when name contains the internal pattern '._'
	for i := 0; i < len(name)-1; i++ {
		if name[i] == '.' && name[i+1] == '_' {
			return true
		}
	}
	return false
}

// GetVars runs the C++ getVar logic asynchronously and returns a channel
// that will receive a slice of tuples (name, value) encoded as two DynamicVar
// elements: [nameAsDynamicVar, value].
func (c *TheController) GetVars(name string, defaultValue misc.DynamicVar) chan GetVarsResult {
	out := make(chan GetVarsResult, 1)
	go func() {
		// readFromDb recursively collects values
		var readFromDb func(string, bool) []misc.Tuple[misc.DynamicVar]
		readFromDb = func(nname string, childsToo bool) []misc.Tuple[misc.DynamicVar] {
			res := []misc.Tuple[misc.DynamicVar]{}
			vh := NewControllerVarHelper(c.db)
			vh.SetControlledVarName(nname)

			dv := vh.GetValue()
			if (&dv).GetString() != "" {
				res = append(res, misc.NewTuple[misc.DynamicVar](misc.NewDynamicVar(nname), dv))
			}

			if childsToo {
				prefix := nname
				if prefix != "" {
					prefix = prefix + "."
				}
				childs := vh.GetChildNames()
				for _, curr := range childs {
					tmp := readFromDb(prefix+curr, childsToo)
					// insert at beginning to match original order
					if len(tmp) > 0 {
						res = append(tmp, res...)
					}
				}
			}
			return res
		}

		// validate
		if name == "" {
			out <- GetVarsResult{nil, errors.New("variable name cannot be empty")}
			return
		}

		childsToo := false
		namep := name
		if strings.Contains(namep, "*") {
			if strings.HasSuffix(namep, "*") {
				childsToo = true
				// strip trailing '*' and possible trailing '.'
				if strings.HasSuffix(namep, ".*") {
					namep = namep[:len(namep)-2]
				} else {
					namep = namep[:len(namep)-1]
				}

				namep = strings.TrimSuffix(namep, ".")
			} else {
				// invalid wildcard position
				out <- GetVarsResult{nil, errors.New("wildcard can only be used at the end of the var name")}
				return
			}
		}

		values := readFromDb(namep, childsToo)
		if len(values) == 0 {
			values = append(values, misc.NewTuple[misc.DynamicVar](misc.NewDynamicVar(namep), defaultValue))
		}
		out <- GetVarsResult{values, nil}
	}()
	return out
}

// DelVar removes variables matching the name (possibly wildcard) asynchronously.
func (c *TheController) DelVar(varname string) chan error {
	out := make(chan error, 1)
	go func() {
		resChan := c.GetVars(varname, misc.NewDynamicVar(""))
		valRes := <-resChan
		if valRes.Err != nil {
			out <- valRes.Err
			return
		}
		vals := valRes.Values
		if len(vals) == 0 {
			out <- errors.New("error deleting variable: maybe it doesn't exist")
			return
		}

		// delete each found var
		for _, t := range vals {
			nameDV := t.At(0)
			dv := nameDV
			nameStr := (&dv).GetString()
			// delete from storage under "vars." prefix
			c.db.DeleteValue("vars."+nameStr, false)
			go c.notifyVarModification(nameStr, misc.NewDynamicVar(""))
		}
		out <- nil
	}()
	return out
}

// GetChildsOfVar returns immediate child names for parentName asynchronously.
func (c *TheController) GetChildsOfVar(parentName string) chan []string {
	out := make(chan []string, 1)
	go func() {
		if strings.Contains(parentName, "*") {
			out <- []string{}
			return
		}
		vh := NewControllerVarHelper(c.db)
		vh.SetControlledVarName(parentName)
		out <- vh.GetChildNames()
	}()
	return out
}

// LockVar tries to acquire a lock for a variable with a timeout in milliseconds.
// Runs in background and returns a channel with the resulting error (nil on success).
func (c *TheController) LockVar(varName string, timeoutMs uint) chan error {
	ch := make(chan error, 1)
	go func() {
		vh := NewControllerVarHelper(c.db)
		vh.SetControlledVarName(varName)
		ch <- vh.Lock(timeoutMs)
	}()
	return ch
}

// UnlockVar releases the lock for a variable. Runs in background.
func (c *TheController) UnlockVar(varName string) chan error {
	ch := make(chan error, 1)
	go func() {
		vh := NewControllerVarHelper(c.db)
		vh.SetControlledVarName(varName)
		ch <- vh.Unlock()
	}()
	return ch
}

// IsVarLocked checks if a variable is locked. Runs in background and returns a channel with the boolean result.
func (c *TheController) IsVarLocked(varName string) chan bool {
	ch := make(chan bool, 1)
	go func() {
		vh := NewControllerVarHelper(c.db)
		vh.SetControlledVarName(varName)
		ch <- vh.IsLocked()
	}()
	return ch
}

// ObserveVar registers a client observation of a variable and sends the
// current value to the client via the API. If the client is disconnected,
// the controller will schedule a liveness check.
func (c *TheController) ObserveVar(varName, clientId, customIdsAndMetainfo string, api apis.IApi) {
	vh := NewControllerVarHelper(c.db)
	vh.SetControlledVarName(varName)
	if !vh.IsClientObserving(clientId, customIdsAndMetainfo) {
		client := NewControllerClientHelper(c.db, clientId, api)
		vh.RegisterObservation(clientId, customIdsAndMetainfo)
		client.RegisterNewObservation(varName)

		// send current var
		vals := c.getVarInternal(varName)
		payload := []misc.Tuple[string]{}
		for _, dv := range vals {
			// dv: tuple(name, value)
			payload = append(payload, misc.NewTuple[string](dv.At(0), customIdsAndMetainfo, dv.At(1)))
		}
		dv := client.Notify(payload)
		if (&dv).GetString() != "LIVE" {
			c.checkClientLiveTime(*client)
		}
	}
}

// StopObservingVar removes a client's observation for a variable.
func (c *TheController) StopObservingVar(varName, clientId, customIdsAndMetainfo string, api apis.IApi) {
	vh := NewControllerVarHelper(c.db)
	vh.SetControlledVarName(varName)
	vh.RemoveObservationByClientAndMetadata(clientId, customIdsAndMetainfo)
	ch := NewControllerClientHelper(c.db, clientId, api)
	ch.UnregisterObservation(varName)
}

// ApiStarted registers a new API implementation with the controller.
func (c *TheController) ApiStarted(api apis.IApi) {
	c.apis[api.GetApiId()] = api
}

// ClientConnected registers a client and updates it with currently observed vars.
func (c *TheController) ClientConnected(clientId string, api apis.IApi) (string, int) {
	if clientId == "" {
		clientId = strconv.FormatInt(time.Now().UnixNano(), 10)
	}
	client := NewControllerClientHelper(c.db, clientId, api)
	observing := client.GetObservingVarsCount()
	go c.updateClientAboutObservatingVars(*client)
	return clientId, observing
}

// updateClientAboutObservatingVars sends current values for all variables the client observes.
func (c *TheController) updateClientAboutObservatingVars(ch ControllerClientHelper) {
	observingVars := ch.GetObservingVars()
	for _, v := range observingVars {
		vals := c.getVarInternal(v)
		payload := []misc.Tuple[string]{}
		for _, dv := range vals {
			vName := dv.At(0)
			vValue := dv.At(1)
			varHelper := NewControllerVarHelper(c.db)
			varHelper.SetControlledVarName(vName)

			metadatas := varHelper.GetMetadatasOfAClient(ch.GetClientId())
			if len(metadatas) == 0 {
				payload = append(payload, misc.NewTuple[string](vName, "", vValue))
			} else {
				for _, metadata := range metadatas {
					payload = append(payload, misc.NewTuple[string](vName, metadata, vValue))
				}
			}
		}
		dv := ch.Notify(payload)
		if (&dv).GetString() != "LIVE" {
			c.checkClientLiveTime(ch)
			return
		}
	}
}

// checkClientLiveTime checks and deletes a client when it has been disconnected
// for longer than the configured timeout.
func (c *TheController) checkClientLiveTime(ch ControllerClientHelper) {
	if !ch.IsConnected() {
		if ch.TimeSinceLastLiveTime() >= c.maxTimeWaitingClientSecond {
			c.deleteClient(ch)
		}
	}
}

// deleteClient removes a client from observation and removes its observing entries.
func (c *TheController) deleteClient(ch ControllerClientHelper) {
	vars := ch.GetObservingVars()
	ch.RemoveClientFromObservationSystem()
	for _, currVar := range vars {
		vh := NewControllerVarHelper(c.db)
		vh.SetControlledVarName(currVar)
		vh.RemoveAllClientObservations(ch.GetClientId())
	}
}

// notifyVarModification informs clients about a variable change.
func (c *TheController) notifyVarModification(varName string, value misc.DynamicVar) {
	go func() {
		vh := NewControllerVarHelper(c.db)
		vh.SetControlledVarName(varName)
		obs := vh.GetAllObservations()
		c.notifyClientsAboutVarChange(obs, varName, value)
		c.notifyParentGenericObservers(varName, varName, value)

		// also notify wildcard children
		childWildcard := NewControllerVarHelper(c.db)
		childWildcard.SetControlledVarName(varName + ".*")
		c.notifyClientsAboutVarChange(childWildcard.GetAllObservations(), varName, value)
	}()
}

// notifyParentGenericObservers notifies parents that have generic observers (a.* patterns)
func (c *TheController) notifyParentGenericObservers(varName, changedVarName string, value misc.DynamicVar) {
	if idx := lastDotIndex(varName); idx != -1 {
		parent := varName[:idx]
		vh := NewControllerVarHelper(c.db)
		vh.SetControlledVarName(parent + ".*")
		c.notifyClientsAboutVarChange(vh.GetAllObservations(), changedVarName, value)
		c.notifyParentGenericObservers(parent, changedVarName, value)
	}
}

func lastDotIndex(s string) int {
	for i := len(s) - 1; i >= 0; i-- {
		if s[i] == '.' {
			return i
		}
	}
	return -1
}

// notifyClientsAboutVarChange sends change notifications to observing clients.
func (c *TheController) notifyClientsAboutVarChange(observations map[ObservationID]misc.Tuple[string], changedVarName string, value misc.DynamicVar) {
	// idxs := []int{}
	// idxToKey := map[int]ObservationID{}
	// for k := range observations {
	// if i, err := strconv.Atoi(string(k)); err == nil {
	// idxs = append(idxs, i)
	// idxToKey[i] = k
	// }
	// }
	// sort.Sort(sort.Reverse(sort.IntSlice(idxs)))

	// for _, i := range idxs {
	for key, _ := range observations {
		tup := observations[key]

		//tup := observations[idxToKey[i]]
		clientId := tup.At(0)
		metadata := tup.At(1)

		go func(clientId, metadata string) {
			ch, err := NewControllerClientHelperWithApis(c.db, clientId, c.apis)
			if err != nil {
				if err == ErrAPIIdNotFound {
					c.log.Error("TheController", "Client notification failed: API not found for client "+clientId)
					ch2 := NewControllerClientHelper(c.db, clientId, nil)
					c.deleteClient(*ch2)
					return
				}
				c.log.Error("TheController", "Client notification failed: unknown error")
				return
			}

			payload := []misc.Tuple[string]{misc.NewTuple[string](changedVarName, metadata, (&value).GetString())}
			res := ch.Notify(payload)
			if (&res).GetString() != "LIVE" {
				c.checkClientLiveTime(*ch)
			}
		}(clientId, metadata)
	}
}

// getVarInternal reads a variable and its children and returns a slice of tuples (name, valueString)
func (c *TheController) getVarInternal(name string) []misc.Tuple[string] {
	res := []misc.Tuple[string]{}
	vh := NewControllerVarHelper(c.db)
	vh.SetControlledVarName(name)
	dv := vh.GetValue()
	if (&dv).GetString() != "" {
		res = append(res, misc.NewTuple[string](name, (&dv).GetString()))
	}
	childs := vh.GetChildNames()
	for _, ch := range childs {
		full := name
		if full != "" {
			full += "." + ch
		} else {
			full = ch
		}
		vvh := NewControllerVarHelper(c.db)
		vvh.SetControlledVarName(full)
		dv2 := vvh.GetValue()
		res = append(res, misc.NewTuple[string](full, (&dv2).GetString()))
	}
	return res
}

// GetSystemVersion returns the configured system version string.
func (c *TheController) GetSystemVersion() string {
	return c.systemVersion
}
